account_fiscal_product_rule/models/account_fiscal_position.py

```diff
@@ -0,0 +1,77 @@
+# Copyright 2022 Akretion France (http://www.akretion.com)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class AccountFiscalPositionProductRule(models.Model):
+    _name = "account.fiscal.position.product.rule"
+    _description = "Account Fiscal Position Rule in Product"
+
+    name = fields.Char(required=True)
+    fiscal_position_id = fields.Many2one("account.fiscal.position", required=True)
+    product_tmpl_ids = fields.Many2many("product.template")
+    product_category_ids = fields.Many2many("product.category")
+    account_income_id = fields.Many2one("account.account")
+    account_expense_id = fields.Many2one("account.account")
+    seller_tax_ids = fields.Many2many("account.tax", "account_tax_seller")
+    supplier_tax_ids = fields.Many2many(
+        "account.tax",
+        "account_tax_supplier",
+    )
+    company_id = fields.Many2one(related="fiscal_position_id.company_id")
+
+    @api.constrains("product_tmpl_ids")
+    def _check_no_duplicate_fiscal_position_on_product(self):
+        self.product_tmpl_ids._check_no_duplicate_fiscal_position()
+
+    @api.constrains("product_category_ids")
+    def _check_no_duplicate_fiscal_position_on_category(self):
+        self.product_category_ids._check_no_duplicate_fiscal_position()
+
+
+class ProductRuleMixin(models.AbstractModel):
+    _name = "product.rule.mixin"
+    _description = "Product Rule Mixin"
+
+    fiscal_position_product_rule_ids = fields.Many2many(
+        "account.fiscal.position.product.rule", string="Product Fiscal Rules"
+    )
+
+    @api.constrains("fiscal_position_product_rule_ids")
+    def _check_no_duplicate_fiscal_position(self):
+        for record in self:
+            fps = []
+            for rule in record.fiscal_position_product_rule_ids:
+                if rule.fiscal_position_id in fps:
+                    raise ValidationError(
+                        _(
+                            "A Fiscal Position Product Rule already exists for this product "
+                            "or product category with this fiscal position !"
+                        )
+                    )
+                else:
+                    fps.append(rule.fiscal_position_id)
+
+
+class AccountFiscalPosition(models.Model):
+    _inherit = "account.fiscal.position"
+
+    fiscal_position_product_rule_ids = fields.One2many(
+        "account.fiscal.position.product.rule",
+        "fiscal_position_id",
+        string="Product Fiscal Rules",
+    )
+
+    def map_tax(self, taxes, product=None, partner=None):
```

Comment:
As the rule is not dependent anymore from the (product) source tax, what about price included or not?
I think you need to ensure the new tax has the same `price_include` state